### ISSUE TITLE
COCO CRA task

### ISSUE THREAD

**Issue by Arsiuuu** [microsoft/LLM2CLIP#3]
Thanks for the good work! I want to determine some details of  COCO CRA task:
1. captions_val2014.json has over 40000 images, are they all be used for CRA?
2. Is the Top-1 accuracy means the Recall@1 of the first caption to the second caption?
![Image](https://github.com/user-attachments/assets/80f1e8ba-add7-4bb9-a0f0-035d70747e07)

**Comment by Yif-Yang** [comment]
1. Yes.

1. Your understanding is correct. This fully demonstrates that the original LLM cannot identify different descriptions of the same image from a set of candidates. However, after fine-tuning with caption2caption, it can effectively distinguish them.

Thank you for recognizing our work~

**Comment by Arsiuuu** [comment]
I was frustrated because I measured a very low R@1(around 10) of CLIP-L-14. Feel it is a bit difficult to maintain 60-70 R@1 for CRA considering 40,504 caption candidates for 1 query. Maybe there is something wrong with the dataset I compiled from [this](https://github.com/tylin/coco-caption/blob/master/annotations/captions_val2014.json). Can you share the test set used for CRA if convenient? Thank you~
![Image](https://github.com/user-attachments/assets/100d285f-1a99-42b7-b6b6-cdfa61721ef3)

```
model = CLIPTextModelWithProjection.from_pretrained("ckpt/clip-vit-large-patch14").to(device)
tokenizer = AutoTokenizer.from_pretrained("/ckpt/clip-vit-large-patch14")
model.eval()
print("model done!")

text_list_1 = []
text_list = []

with torch.no_grad():
    for i, (caption_1, caption) in enumerate(tqdm(dataset)):
        inputs = tokenizer([caption_1, caption], padding=True, return_tensors="pt").to(device)
        outputs = model(**inputs)
        text_embeds = outputs.text_embeds

        text_list.append(text_embeds[1].unsqueeze(0))
        text_list_1.append(text_embeds[0].unsqueeze(0))

text_features = torch.cat(text_list, dim=0)
text_features = text_features / text_features.norm(dim=-1, keepdim=True)

text_features_1 = torch.cat(text_list_1, dim=0)
text_features_1 = text_features_1 / text_features_1.norm(dim=-1, keepdim=True)
    
def compute_metrics(x):
    try:
        x = x.cpu().numpy()
    except:
        pass
    sx = np.sort(-x, axis=1)
    d = np.diag(-x)
    d = d[:, np.newaxis]
    ind = sx - d
    ind = np.where(ind == 0)
    ind = ind[1]
    metrics = {}
    metrics['R1'] = float(np.sum(ind == 0)) * 100 / len(ind)
    metrics['R5'] = float(np.sum(ind < 5)) * 100 / len(ind)
    metrics['R10'] = float(np.sum(ind < 10)) * 100 / len(ind)
    metrics['MR'] = np.median(ind) + 1
    metrics["MedianR"] = metrics['MR']
    metrics["MeanR"] = np.mean(ind) + 1
    return metrics

temp = 100.
sim_matrix = (text_features_1 @ text_features.T) * temp
tv_metrics = compute_metrics(sim_matrix)
vt_metrics = compute_metrics(sim_matrix.T)
print('s2l_metrics: ',  tv_metrics)
print('l2s_metrics: ',  vt_metrics)
```

**Comment by Yif-Yang** [comment]
I reconfirmed with the team members, and it turns out that we are actually using the 5K test set of COCO. This is our mistake, as we mistakenly thought it was called the validation set. We'll correct this oversight. Could @Arsiuuu  please try again using the 5K test set of COCO? Thank you for pointing this out, and we apologize for the confusion.

**Comment by Arsiuuu** [comment]
I use the 5000 split https://huggingface.co/datasets/nlphuji/mscoco_2014_5k_test_image_text_retrieval/resolve/main/test_5k_mscoco_2014.csv
and extract first two captions of every image. The result is 23.40 for CLIPTextModelWithProjection and 25.14 for CLIPTextModel😢

**Comment by Yif-Yang** [comment]
> I use the 5000 split https://huggingface.co/datasets/nlphuji/mscoco_2014_5k_test_image_text_retrieval/resolve/main/test_5k_mscoco_2014.csv and extract first two captions of every image. The result is 23.40 for CLIPTextModelWithProjection and 25.14 for CLIPTextModel😢

We are now trying to reproduce for you, should be something wrong here. will inform you when we know why

**Comment by wswaq** [comment]
> I use the 5000 split https://huggingface.co/datasets/nlphuji/mscoco_2014_5k_test_image_text_retrieval/resolve/main/test_5k_mscoco_2014.csv and extract first two captions of every image. The result is 23.40 for CLIPTextModelWithProjection and 25.14 for CLIPTextModel😢

Sorry for the confusion, it was a bug during our testing. We're correcting it now.
Here are more details: our evaluation uses a batch size of 128 (i.e., a 128*128 distance matrix), and we compute the average over the entire dataset. For the CLIP text model, we use its pooler_output; for EVA CLIP, we use encode_text; and for LLaMA3, we use its last_hidden_state.
I also provide the results of directly computing over the entire dataset here:
| Language Model            | CRA   |
|-----------------|--------|
| CLIP-L/14       | 25.32  |
| EVA02-L/14      | 27.14  |
| Llama3-8B       | 5.22   |
| Llama3.2-1B     | 5.66   |
| Llama3-8B-CC    | 29.46  |
| Llama3.2-1B-CC  | 29.42  |

**Comment by Arsiuuu** [comment]
So the CRA result in the article is an average of only 128 candidates each time, while the version provided here is a version of 5000 candidates. Is my understanding correct?

**Comment by wswaq** [comment]
> So the CRA result in the article is an average of only 128 candidates each time, while the chart provided here is a version of 5000 candidates. Is my understanding correct?

Yes, your understanding is correct.

**Comment by Arsiuuu** [comment]
Thanks for your patience in answering!
